aquadoggo/src/db/stores/document.rs

```diff
@@ -0,0 +1,331 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use std::collections::btree_map::Iter;
+
+use async_trait::async_trait;
+use futures::future::try_join_all;
+use p2panda_rs::document::{DocumentView, DocumentViewHash, DocumentViewId};
+use p2panda_rs::operation::OperationValue;
+use p2panda_rs::schema::SchemaId;
+use sqlx::{query, query_as};
+
+use crate::db::errors::DocumentViewStorageError;
+use crate::db::models::operation::OperationFieldRow;
+use crate::db::provider::SqlStorage;
+use crate::db::traits::{
+    AsStorageDocumentView, DocumentStore, DocumentViewFields, FieldIds, FieldName,
+};
+use crate::db::utils::parse_operation_fields;
+
+/// Aquadoggo struct which will implement AsStorageDocumentView trait.
+#[derive(Debug, Clone)]
+pub struct DocumentViewStorage {
+    document_view: DocumentView,
+    field_ids: FieldIds,
+    schema_id: SchemaId,
+}
+
+impl DocumentViewStorage {
+    pub fn new(document_view: &DocumentView, field_ids: &FieldIds, schema_id: &SchemaId) -> Self {
+        Self {
+            document_view: document_view.clone(),
+            field_ids: field_ids.clone(),
+            schema_id: schema_id.clone(),
+        }
+    }
+}
+
+impl AsStorageDocumentView for DocumentViewStorage {
+    type AsStorageDocumentViewError = DocumentViewStorageError;
+
+    fn id(&self) -> DocumentViewId {
+        self.document_view.id().clone()
+    }
+
+    fn iter(&self) -> Iter<FieldName, OperationValue> {
+        self.document_view.iter()
+    }
+
+    fn get(&self, key: &str) -> Option<&OperationValue> {
+        self.document_view.get(key)
+    }
+
+    fn schema_id(&self) -> SchemaId {
+        self.schema_id.clone()
+    }
+
+    fn field_ids(&self) -> FieldIds {
+        self.field_ids.clone()
+    }
+}
+
+#[async_trait]
+impl DocumentStore<DocumentViewStorage> for SqlStorage {
+    /// Insert a document_view into the db. Requires that all relevent operations are already in
+    /// the db as this method only creates relations between document view fields and their current
+    /// values (last updated operation value).
+    ///
+    /// QUESTION: Is this too implementation specific? It assumes quite a lot about the db
```

Comment:
Does this need to be implemented for `p2panda_rs`? I agree, this implementation is too specific for something else than aquadoggo SQL, but I'm sure we can a) either find something else for a more "generic" p2panda_rs version b) or just leave it unimplemented?

Reply:
We need an `insert_document_view()` method in `p2panda_rs`, but exposing the details of how we're doing that is too specific to our db layout, yeh. I would say that we should develop `DocumentView` to contain the data we need though, as it's useful and not too much overhead, then the signature for this method could be something like:

```
    async fn insert_document_view(
        &self,
        document_view: &DocumentView,
        schema_id: &SchemaId,
    ) -> Result<bool, DocumentViewStorageError>;
 ```
 
 I'll open an issue for changes in `DocumentView`.